api/src/utils/redirection.ts

```diff
@@ -0,0 +1,139 @@
+import { FastifyRequest } from 'fastify';
+import jwt from 'jsonwebtoken';
+
+// import { allowedOrigins } from '../../config/allowed-origins';
+import { availableLangs } from '../../../shared/config/i18n';
+import { allowedOrigins } from './allowed-origins';
+
+// process.env.HOME_LOCATION is being used as a fallback here. If the one
+// provided by the client is invalid we default to this.
+import { HOME_LOCATION } from './env';
+
+/**
+ * Get the returnTo value.
+ *
+ * @param encryptedParams - The encrypted parameters.
+ * @param secret - The secret key.
+ * @param _homeLocation - The home location.
+ * @returns The returnTo value.
+ */
+export function getReturnTo(
+  encryptedParams: string,
+  secret: jwt.Secret,
+  _homeLocation = process.env.HOME_LOCATION
+) {
+  let params;
+  try {
+    params = jwt.verify(encryptedParams, secret);
+  } catch (e) {
+    // TODO: report to Sentry? Probably not. Remove entirely?
+    console.log(e);
+    // something went wrong, use default params
+    params = {
+      returnTo: `${_homeLocation}/learn`,
+      origin: _homeLocation,
+      pathPrefix: ''
+    };
+  }
+
+  // @ts-expect-error - I'm working on it...
+  return normalizeParams(params, _homeLocation);
+}
+
+/**
+ * Normalize the parameters, making they're valid.
+ *
+ * @param arg - The parameters to normalize.
+ * @param arg.returnTo - The returnTo value.
+ * @param arg.origin - The origin value.
+ * @param arg.pathPrefix - The pathPrefix value.
+ * @param _homeLocation - The home location.
+ * @returns The normalized parameters.
+ */
+export function normalizeParams(
+  {
+    returnTo,
+    origin,
+    pathPrefix
+  }: { returnTo?: string; origin?: string; pathPrefix?: string },
+  _homeLocation = HOME_LOCATION
+) {
+  // coerce to strings, just in case something weird and nefarious is happening
+  // TODO: validate, don't coerce
+  returnTo = '' + returnTo;
+  origin = '' + origin;
+  pathPrefix = '' + pathPrefix;
+  // we add the '/' to prevent returns to
+  // www.freecodecamp.org.somewhere.else.com
```

Comment:
Do we have a test for this? In case someone nukes the comments or normalizes the slash in a later refactoring.